### PR TITLE
Chongyangbai/kvpair grounding

### PR DESCRIPTION
**Pull request by cy-bai**
Update key value pair schema with optional grounding information associated with each key.

### CODE REVIEW

File: COCO_DATA_FORMAT.md [hunk 1]
```diff
@@ -236,7 +236,7 @@ It is a generic variation of an image-text dataset, where the input consists of
   - images, image is optionally associated with a metadata dictionary which stores the text attributes of interest for the image. For example, image is a product catalog image: `{'metadata': {'catalog': true}}`, capture location of an image: `{'metadata': {'location': 'street'}}`, information of the assembly component captured in image of a defect detection dataset: `{'metadata': {'name': 'Hex Head Lag Screw', 'type': '3/8-inch x 4-inch'}}`  
   - text (optional), a dictionary with keys being field names e.g. `{'text': {'question': 'a specific question related to the images input'}}`
 - output:
-  - key-value pairs, a dictionary with keys being the attributes of interest
+  - key-value pairs, a dictionary with keys being the attributes of interest, values being dictionary that stores the actual attribute value in "value" and optionally a list of grounded bboxes in "groundings". "groundings" are for single-image annotation only. Each bbox follows [BBox Format](#bbox-format).
```
Comment (juliagongms) [inline COCO_DATA_FORMAT.md]: Nit: `values being dictionaries that store`

File: COCO_DATA_FORMAT.md [hunk 2]
```diff
@@ -252,3 +252,32 @@ The dataset format is a simple variation of COCO, where `image_id` of an annotat
...
+            "height": 224,
+            "file_name": "1.jpg",
+            "zip_file": "test.zip"
+        }
+    ],
+    "annotations": [
+        {
+            "id": 1,
+            "image_ids": [1],
+            "key_value_pairs": {
```
Comment (juliagongms) [inline COCO_DATA_FORMAT.md]: Just noticed this, should we rename this to match the API terminology (`fieldSchema` for the schema and `fields` in the response)?

File: COCO_DATA_FORMAT.md [hunk 3]
```diff
@@ -236,7 +236,7 @@ It is a generic variation of an image-text dataset, where the input consists of
...
   - text (optional), a dictionary with keys being field names e.g. `{'text': {'question': 'a specific question related to the images input'}}`
 - output:
-  - key-value pairs, a dictionary with keys being the attributes of interest
+  - key-value pairs, a dictionary with keys being the attributes of interest, values being dictionary that stores the actual attribute value in "value" and optionally a list of grounded bboxes in "groundings". "groundings" are for single-image annotation only. Each bbox follows [BBox Format](#bbox-format).
 
 The dataset format is a simple variation of COCO, where `image_id` of an annotation entry is replaced with `image_ids` to support multi-image annotation. In each annotation entry, `key_value_pairs` field is required, `text` is optional. In each image entry, `metadata` is optional. Below is an example of multi-image question answering.
```
Comment (juliagongms) [inline COCO_DATA_FORMAT.md]: Should rename `key_value_pairs` to `fields` here as well if we change it in the annotations.

File: DATA_PREPARATION.md [hunk 4]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
-| items       | FieldSchema                       | define the FieldSchema for all items in array                                                                                                                  | yes when type is array  |
-
-More details can be found at [`vision-datasets/vision_datasets/key_value_pair/manifest.py`](vision_datasets/key_value_pair/manifest.py)
+| Property         | Type                      | Example                                                                                                     | Required?               |
+| :--------------- | :------------------------ | :---------------------------------------------------------------------------------------------------------- | :---------------------- |
+| type             | FieldValueType            | JSON type: string, number, integer, boolean, array, object.                                                 | yes                     |
+| description      | string                    | answer to question                                                                                          | no                      |
```
Comment (juliagongms) [inline DATA_PREPARATION.md]: Nit: Instead of "answer to question", maybe "describes the field in more detail"?

File: DATA_PREPARATION.md [hunk 5]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
-| enum        | list[string \| number \| integer] | only work when type is string or number or integer                                                                                                             | no                      |
-| properties  | dict[string, FieldSchema]         | define FieldSchema of each subfield                                                                                                                            | yes when type is object |
-| items       | FieldSchema                       | define the FieldSchema for all items in array                                                                                                                  | yes when type is array  |
-
-More details can be found at [`vision-datasets/vision_datasets/key_value_pair/manifest.py`](vision_datasets/key_value_pair/manifest.py)
+| Property         | Type                      | Example                                                                                                     | Required?               |
```
Comment (juliagongms) [inline DATA_PREPARATION.md]: Since these are all descriptions, maybe change `Example` to `Details`?

File: DATA_PREPARATION.md [hunk 6]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
+| :--------------- | :------------------------ | :---------------------------------------------------------------------------------------------------------- | :---------------------- |
+| type             | FieldValueType            | JSON type: string, number, integer, boolean, array, object.                                                 | yes                     |
+| description      | string                    | answer to question                                                                                          | no                      |
+| examples         | list[string]              | examples of field content                                                                                   | no                      |
+| classes          | dict[str, ClassSchema]    | dictionary that maps each class name to `ClassSchema`.                                                      | no                      |
+| properties       | dict[string, FieldSchema] | define FieldSchema of each subfield                                                                         | yes when type is object |
```
Comment (juliagongms) [inline DATA_PREPARATION.md]: Nit: "defines"

File: DATA_PREPARATION.md [hunk 7]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
+| items            | FieldSchema               | define the FieldSchema for all items in array                                                               | yes when type is array  |
```
Comment (juliagongms) [inline DATA_PREPARATION.md]: Nit: "defines"

File: DATA_PREPARATION.md [hunk 8]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
+| includeGrounding | boolean                   | Whether annotation of this field contains bbox groundings, if true, bboxes are stored in `groundings` field | No, default false       |
```
Comment (juliagongms) [inline DATA_PREPARATION.md]: Nit: lowercase "whether" to follow the convention of the other fields?
Also, grammar/clarity, maybe: "whether annotation of this field should contain bbox groundings; if true, bboxes are stored in the `groundings` field of the annotation"

File: vision_datasets/key_value_pair/manifest.py [hunk 9]
```diff
@@ -10,47 +10,66 @@ class KeyValuePairValueTypes(Enum):
...
                  items: 'KeyValuePairFieldSchema' = None,
-                 properties: Dict[str, 'KeyValuePairFieldSchema'] = None) -> None:
+                 properties: Dict[str, 'KeyValuePairFieldSchema'] = None,
+                 includeGrounding: bool = False) -> None:
         """
         Key-value pair schema for each field.
-        
+        The annotation of each key is a dictionary containing
+            1. "value"  field that contains the annotation
+            2. "groundings" field enabled when includeGrounding=True, which contains a list of grounded bounding boxes in image for the annotation
```
Comment (juliagongms) [inline vision_datasets/key_value_pair/manifest.py]: Nit: "in the image"

File: vision_datasets/key_value_pair/manifest.py [hunk 10]
```diff
@@ -10,47 +10,66 @@ class KeyValuePairValueTypes(Enum):
...
+
         Args:
             type (str): type of the field, one of KeyValuePairValueTypes names in lower case.
             description (str): description of the field
             examples (list of str): examples of the field
-            enum (list of str/float/int): if the field is restricted to a list of values, define that list. Only work when type is string/number/integer.
+            classes (dict[str, KeyValuePairClassSchema]): if the field is restricted to a list of classes, define the map from class name to its information. Only work when type is string.
```
Comment (juliagongms) [inline vision_datasets/key_value_pair/manifest.py]: Nit: "defines", "works"

Comment (juliagongms) [inline vision_datasets/key_value_pair/manifest.py]: Ah, I think this was left out of the spec document. I think we will continue to support classes for number and integer, not just string. Made the change in the schema definition file - can we update it here too?

File: vision_datasets/key_value_pair/manifest.py [hunk 11]
```diff
@@ -59,17 +78,19 @@ def __eq__(self, other) -> bool:
...
     
     def _check(self):
         if self.type not in self.TYPE_NAME_TO_PYTHON_TYPE:
             raise ValueError(f'Invalid type: {self.type}')
-        if self.enum and self.type not in [KeyValuePairValueTypes.STRING, KeyValuePairValueTypes.NUMBER, KeyValuePairValueTypes.INTEGER]:
-            raise ValueError('enum is only allowed for string, number, integer types')
-        if self.enum and len(self.enum) != len(set(self.enum)):
-            raise ValueError('enum values must be unique')
+        if self.classes:
+            if self.type != KeyValuePairValueTypes.STRING:
```
Comment (juliagongms) [inline vision_datasets/key_value_pair/manifest.py]: Can update these checks to use string | number | integer

File: vision_datasets/key_value_pair/manifest.py [hunk 12]
```diff
@@ -81,19 +102,36 @@ def __init__(self, name: str, field_schema_dict: Dict, description: str = None)
...
         return isinstance(other, KeyValuePairSchema) and self.name == other.name and self.field_schema == other.field_schema and self.description == other.description
 
 
 class KeyValuePairLabelManifest(MultiImageLabelManifest):
     """
+    Label manifest for key-value pair annotations. The "key_value_pairs" field follows KeyValuePairSchema.
+    For example, the label data can be:   
     {
-        "key_value_pairs": {"key1": "value1", ...},
+        "key_value_pairs": {
```
Comment (juliagongms) [inline vision_datasets/key_value_pair/manifest.py]: If we rename this key, we should rename this to `fields`

File: vision_datasets/key_value_pair/manifest.py [hunk 13]
```diff
@@ -81,19 +102,36 @@ def __init__(self, name: str, field_schema_dict: Dict, description: str = None)
...
+    {
+        "name": "example key value pair schema",
+        "fieldSchema": {
+            "key1": {"type": "string", "includeGrounding": true},
+            "key2": {"type": "string"},
+            ...
+        }
     }
     """
     LABEL_KEY = 'key_value_pairs'
```
Comment (juliagongms) [inline vision_datasets/key_value_pair/manifest.py]: Need to change to `fields` if we change the name

File: DATA_PREPARATION.md [hunk 14]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
+
+Definition of `ClassSchema`:
+| Property    | Type   | Example                          | Required?         |
```
Comment (juliagongms) [inline DATA_PREPARATION.md]: For consistency, maybe `Example` -> `Details` and change the content of the first row to `describes the class in more detail, e.g., "long, thin, surface-level mark"`?

File: vision_datasets/key_value_pair/manifest.py (hunk 9 above)
Comment (cy-bai) [reply vision_datasets/key_value_pair/manifest.py]: thanks for the help!

**Comment by juliagongms** [comment]
Looks good to me, thanks Chongyang!

**Review by juliagongms — APPROVED** [review]
Looks good to me with these changes, thanks Chongyang!

File: DATA_PREPARATION.md [hunk 15]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
 
 The `fieldSchema` is a dictionary mapping each field to a dictionary called `FieldSchema`. In above example, there are two fields: `answer` and `rationale`. Below is the definition of `FieldSchema`:
```
Comment (pj-ms) [inline DATA_PREPARATION.md]: "The `fieldSchema` is a dictionary mapping each field to a dictionary called `FieldSchema`"

Wording a little bit confusing

File: DATA_PREPARATION.md [hunk 16]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
-| enum        | list[string \| number \| integer] | only work when type is string or number or integer                                                                                                             | no                      |
-| properties  | dict[string, FieldSchema]         | define FieldSchema of each subfield                                                                                                                            | yes when type is object |
-| items       | FieldSchema                       | define the FieldSchema for all items in array                                                                                                                  | yes when type is array  |
-
-More details can be found at [`vision-datasets/vision_datasets/key_value_pair/manifest.py`](vision_datasets/key_value_pair/manifest.py)
+| Property         | Type                      | Details                                                                                                                                 | Required?               |
```
Comment (pj-ms) [inline DATA_PREPARATION.md]: those format info should be part of COCO_DATA_FORMAT.md?

File: DATA_PREPARATION.md [hunk 17]
```diff
@@ -200,15 +200,21 @@ For `key_value_pair` dataset, an additional field `schema` is required to define
...
+| examples         | list[string]              | examples of field content                                                                                                               | no                      |
+| classes          | dict[str, ClassSchema]    | dictionary that maps each class name to `ClassSchema`.                                                                                  | no                      |
+| properties       | dict[string, FieldSchema] | defines FieldSchema of each subfield                                                                                                    | yes when type is object |
+| items            | FieldSchema               | defines the FieldSchema for all items in array                                                                                          | yes when type is array  |
+| includeGrounding | boolean                   | whether annotation of this field should contain bbox groundings; if true, bboxes are stored in the `groundings` field of the annotation | No, default false       |
```
Comment (pj-ms) [inline DATA_PREPARATION.md]: "whether annotation of this field should contain bbox groundings"
=>
"whether this annotation field has bbox groundings associated"

Comment (pj-ms) [inline DATA_PREPARATION.md]: lrwh? ltrb?

Comment (pj-ms) [inline DATA_PREPARATION.md]: add notes about single image restriction?

Can we check the restriction in the code and add a test?

File: DATA_PREPARATION.md (hunk 16 above)
Comment (cy-bai) [reply DATA_PREPARATION.md]: yes, moved there and also refined the doc to be clearer.

File: DATA_PREPARATION.md (hunk 17 above)
Comment (cy-bai) [reply DATA_PREPARATION.md]: good catch, added

Comment (cy-bai) [reply DATA_PREPARATION.md]: moved to COCO_DATA_FORMAT.md, it is the same as how we define bbox_format for OD dataset. Added this to the doc

**Review by pj-ms — APPROVED** [review]

**Pull request merged**